packages/xo-server/src/xapi/mixins/patching.js

```diff
@@ -464,4 +465,97 @@ export default {
     // host-by-host install
     throw new Error('non pool-wide install not implemented')
   },
+
+  async rollingPoolUpdate() {
+    const hosts = filter(this.objects.all, { $type: 'host' })
+    await Promise.all(hosts.map(host => host.$call('assert_can_evacuate')))
+
+    log.debug('Install patches')
+    await this.installPatches()
+
+    // Remember on which hosts the running VMs are
+    const vmsByHost = mapValues(
+      groupBy(
+        filter(this.objects.all, {
+          $type: 'VM',
+          power_state: 'Running',
+          is_control_domain: false,
+        }),
+        vm => {
+          const hostId = vm.$resident_on?.$id
+
+          if (hostId === undefined) {
+            throw new Error('Could not find host of all running VMs')
+          }
+
+          return hostId
+        }
+      ),
+      vms => vms.map(vm => vm.$id)
+    )
+
+    // Put master in first position to restart it first
+    const indexOfMaster = hosts.findIndex(host => host.$ref === this.pool.master)
+    if (indexOfMaster > 0) {
+      ;[hosts[0], hosts[indexOfMaster]] = [hosts[indexOfMaster], hosts[0]]
+    }
+
+    // Restart all the hosts one by one
+    for (const host of hosts) {
+      const hostId = host.uuid
+      const metricsRef = host.metrics
```

Comment:
I don't think refs are guaranteed to survive pool master restart.

Reply:
I tested it and it worked. Do you think there's still a chance that the ref changes?

Reply:
Maybe we should live i like that but add a comment in the code explaining this could be an issue to give pointers in case it happens?